lib/include/tinycrypt/utils.h

```diff
@@ -69,6 +70,24 @@ unsigned int _copy(uint8_t *to, unsigned int to_len,
  */
 void _set(void *to, uint8_t val, unsigned int len);
 
+/**
+ * @brief Set the value 'val' into the buffer 'to', 'len' times, in a way
+ *         which does not risk getting optimized out by the compiler.
+ *
+ * @param to OUT -- destination buffer
+ * @param val IN -- value to be set in 'to'
+ * @param len IN -- number of times the value will be copied
+ */
+#ifdef __GNUC__
+static inline void _set_secure(void *to, uint8_t val, unsigned int len)
+{
+  (void) memset(to, val, len);
+  __asm__ __volatile__("" :: "g"(to) : "memory");
+}
+#else /* ! __GNUC__ */
+extern void _set_secure(void *to, uint8_t val, unsigned int len);
```

Comment:
I would have a default implementation that uses `memset` this way we won't break any application out there. It may be possible that others achieve the same thing using a compiler option. Also, this will break the build, this is built as a static library by default. Though, you can keep a possibility of one bring its own implementation, e.g:
```
#ifdef MY_OWN_SECURE_FUNCTION
extern void _set_secure(void *to, uint8_t val, unsigned int len);
#else
static inline void _set_secure(void *to, uint8_t val, unsigned int len)
{
  (void) memset(to, val, len);
#ifdef __GNUC__
  __asm__ __volatile__("" :: "g"(to) : "memory");
#endif
}
#endif
```

Reply:
Thank you for your comment, and I see your point.
With your suggested method, some applications and builds will end up silently having the memset optimized out, which then degrades the security. In my proposed solution the build would fail, thus making the implementer aware of the special need and also to handle the situation. From a security point of view, I would prefer my solution, but if you do not think it is an issue that the memset could silently be optimized out, I can change to what you suggest. 
What do you think?

I would prefer another define than MY_OWN_SECURE_FUNCTION, though. What about TINYCRYPT_ARCH_HAS_SET_SECURE?

Reply:
That define was just an example :). What I said is that may be cases where you can ask the compiler to NOT optimize out that memset call.
How you are building it ? With your patch this library will not build without change the code when not using GCC.
In other words we are enforcing people to change the build and/or tinycrypt's code to build it.

What I'm suggesting is document this function saying that memset can be optimized-out and this can be a security issue, gives the opportunity to one bring one's function (defining TINYCRYPT_ARCH_HAS_SET_SECURE) or just use a compiler option to not optimize it.

Reply:
Yes, there are of cause cases where the compiler will not optimize out the memset, which depends upon the compiler optimization level setting and which is compiler dependent, hence the need for this abstraction :-)
We are using a propritary compiler, which does not like the __asm__ memory barrier, which is why we are unable to build tinycrypt, without modifications.
My intention was actually to force the user to fix the issue, by implementing their own version of _set_secure, but I see your point in that not being the best solution.

I have updated the PR as you requested. What do you think?